Scripts/XRLineRenderer.cs

```diff
@@ -0,0 +1,424 @@
+    {
+        // See if the data needs initializing
+        if (Initialize())
+        {
+            return;
+        }
+
+        // If it doesn't, go through each point and set the data
+        var pointCounter = 0;
+        var elementCounter = 0;
+        var stepPercent = 0.0f;
+
+        var lastColor = m_Color.Evaluate(stepPercent);
+        m_XRMeshData.SetElementColor(elementCounter, ref lastColor);
+        elementCounter++;
+        pointCounter++;
+        stepPercent += m_StepSize;
+
+        while (pointCounter < m_Positions.Length)
+        {
+            var currentColor = m_Color.Evaluate(stepPercent);
+            m_XRMeshData.SetElementColor(elementCounter, ref lastColor, ref currentColor);
+            elementCounter++;
+
+            m_XRMeshData.SetElementColor(elementCounter, ref currentColor);
+
+            lastColor = currentColor;
+            elementCounter++;
+            pointCounter++;
+            stepPercent += m_StepSize;
+        }
+
+        if (m_Loop)
+        {
+            lastColor = m_Color.Evaluate(stepPercent);
+            m_XRMeshData.SetElementColor(elementCounter, ref lastColor);
+        }
+
+        // Dirty the color meshChain flags so the mesh gets new data
+        m_XRMeshData.SetMeshDataDirty(XRMeshChain.MeshRefreshFlag.Colors);
+        m_MeshNeedsRefreshing = true;
+    }
+
+    protected override void UpdateWidth()
+    {
+        // See if the data needs initializing
+        if (Initialize())
+        {
+            return;
+        }
+
+        // Otherwise, do fast setting
+        var pointCounter = 0;
+        var elementCounter = 0;
+        var stepPercent = 0.0f;
+        
+        // We go through the element list, much like initialization, but only update the width part of the variables
+        var lastWidth = m_WidthCurve.Evaluate(stepPercent) * m_Width;
+        m_XRMeshData.SetElementSize(elementCounter, lastWidth);
+        elementCounter++;
+        pointCounter++;
+
+        stepPercent += m_StepSize;
+
+        while (pointCounter < m_Positions.Length)
+        {
+            var currentWidth = m_WidthCurve.Evaluate(stepPercent) * m_Width;
+
+            m_XRMeshData.SetElementSize(elementCounter, lastWidth, currentWidth);
+            elementCounter++;
+            m_XRMeshData.SetElementSize(elementCounter, currentWidth);
+            lastWidth = currentWidth;
+            elementCounter++;
+            pointCounter++;
+            stepPercent += m_StepSize;
+        }
+
+        if (m_Loop)
+        {
+            var currentWidth = m_WidthCurve.Evaluate(stepPercent) * m_Width;
+            m_XRMeshData.SetElementSize(elementCounter, lastWidth, currentWidth);
+        }
+
+        // Dirty all the VRMeshChain flags so everything gets refreshed
+        m_XRMeshData.SetMeshDataDirty(XRMeshChain.MeshRefreshFlag.Sizes);
+        m_MeshNeedsRefreshing = true;
+    }
+
+    protected override bool Initialize(bool force = false)
```

Comment:
the way this function works / is used is a bit muddled to me.  Looks like it's used both to initialize, as well as to check whether any initialization is needed ?